### PR TITLE
MESH-1620/improve batch functions (EventCounts, ErrorAt)

### PR DESCRIPTION
**Pull request by Neopallium**
### new features

- batch functions return number of events emitted from each call.

### modified API

- Changed `BatchInterrupted`, `BatchOptimisticFailed` and `BatchCompleted` events to include a vector of event counts.
- Wrap the call index and error in a `ErrorAt` type.  This changes `BatchInterrupted` and `BatchOptimisticFailed`.

### CODE REVIEW

File: pallets/runtime/tests/src/utility_test.rs [hunk 1]
```diff
@@ -308,15 +307,22 @@ fn batch_secondary_with_permissions() {
...
 
     // Call the same extrinsics optimistically.
     assert_ok!(Utility::batch_optimistic(bob_origin, calls));
     assert_event_exists!(
-        EventTest::pallet_utility(Event::BatchOptimisticFailed(errors)),
-        errors.len() == 1 && errors[0].0 == 0
+        EventTest::pallet_utility(Event::BatchOptimisticFailed(events, errors)),
+        events.len() == 2
+            && events[0] == 0
+            && events[1] == 1
```
Comment (Centril) [inline pallets/runtime/tests/src/utility_test.rs:323]:
```suggestion
        events == [0, 1]
```

File: pallets/runtime/tests/src/utility_test.rs [hunk 2]
```diff
@@ -308,15 +307,22 @@ fn batch_secondary_with_permissions() {
         )),
     ];
     assert_ok!(Utility::batch(bob_origin.clone(), calls.clone()));
-    assert_event_doesnt_exist!(EventTest::pallet_utility(Event::BatchCompleted));
-    assert_event_exists!(EventTest::pallet_utility(Event::BatchInterrupted(0, _)));
+    assert_event_doesnt_exist!(EventTest::pallet_utility(Event::BatchCompleted(_)));
+    assert_event_exists!(
+        EventTest::pallet_utility(Event::BatchInterrupted(events, err)),
+        events.len() == 1 && events[0] == 0 && err.0 == 0
```
Comment (Centril) [inline pallets/runtime/tests/src/utility_test.rs:313]:
```suggestion
        events == [0] && err.0 == 0
```

File: pallets/utility/src/lib.rs [hunk 3]
```diff
@@ -121,13 +124,15 @@ decl_event! {
...
+        /// Includes a vector of event counts for each dispatch and
+        /// the index of the first failing dispatch as well as the error.
+        BatchInterrupted(EventCounts, ErrorAt),
         /// Batch of dispatches did not complete fully.
-        /// Includes any failed dispatches with their indices and their associated error.
-        BatchOptimisticFailed(Vec<(u32, DispatchError)>),
+        /// Includes a vector of event counts for each call and
+        /// a vector of any failed dispatches with their indices and associated error.
+        BatchOptimisticFailed(EventCounts, Vec<ErrorAt>),
         /// Batch of dispatches completed fully with no error.
```
Comment (Centril) [inline pallets/utility/src/lib.rs:134]:
```suggestion
        /// Batch of dispatches completed fully with no error.
        /// Includes a vector of event counts for each dispatch.
```
Just for consistency

File: pallets/utility/src/lib.rs [hunk 4]
```diff
@@ -320,30 +294,75 @@ decl_module! {
...
+        let mut counts = EventCounts::with_capacity(calls.len());
+        let mut errors = Vec::new();
+        for (index, call) in calls.into_iter().enumerate() {
+            let count = System::<T>::event_count();
+
+            // Dispatch the call in a modified metadata context.
+            let res = Self::dispatch_call(origin.clone(), is_root, call);
+            counts.push(System::<T>::event_count().saturating_sub(count));
+
+            // handle call error
```
Comment (Centril) [inline pallets/utility/src/lib.rs:344]:
```suggestion
            // Handle call error.
```

File: pallets/utility/src/lib.rs [hunk 5]
```diff
@@ -320,30 +294,75 @@ decl_module! {
...
+            if let Err(e) = res {
+                let err_at = (index as u32, e.error);
+                if stop_on_errors {
+                    // Interrupt the batch on first error
```
Comment (Centril) [inline pallets/utility/src/lib.rs:348]:
```suggestion
                    // Interrupt the batch on first error.
```

**Review by Centril — APPROVED** [review]
Some nits, but this LGTM!

File: pallets/runtime/tests/src/utility_test.rs (hunk 1 above)
Comment (Neopallium) [reply pallets/runtime/tests/src/utility_test.rs:323]: I had those on one line, but `cargo fmt` split it over multiple.  I hope it will except that.

Comment (Neopallium) [reply pallets/runtime/tests/src/utility_test.rs:323]: That is cool.  I didn't know that a vec could be compared to a slice like that.

Comment (Centril) [reply pallets/runtime/tests/src/utility_test.rs:323]: Yeah! This impl specifically, https://doc.rust-lang.org/nightly/std/cmp/trait.PartialEq.html#impl-PartialEq%3C%26%27_%20%5BU%5D%3E-1

**Comment by Centril** [comment]
You'll need to add `ErrorAt` and `EventCounts` to the `polymesh_schema.json` to get the integration test to pass (concourse-ci/status).

File: polymesh_schema.json [hunk 6]
```diff
@@ -8,6 +8,8 @@
         },
         "IdentityId": "[u8; 32]",
         "EventDid": "IdentityId",
+				"EventCounts": "Vec<u32>",
+				"ErrorAt": "(u32, DispatchError)",
```
Comment (Centril) [inline polymesh_schema.json:12]:
```suggestion
        "EventDid": "IdentityId",
        "EventCounts": "Vec<u32>",
        "ErrorAt": "(u32, DispatchError)",
```
You might need to fix your editor settings to only use spaces :)

**Comment by Centril** [comment]
Currently waiting on some feedback from SDK folks on which approach works best for them...

**Review by adamdossa — APPROVED** [review]

**Pull request merged**